### PR TITLE
[AutoDiff upstream] define autodiff builtins

### PR DESCRIPTION
**Pull request by marcrasi**
Defines type signatures and SILGen for the following builtins:

```
/// Applies the {jvp|vjp} of `f` to `arg1`, ..., `argN`.
func applyDerivative_arityN_{jvp|vjp}(f, arg1, ..., argN) -> jvp/vjp return type

/// Applies the transpose of `f` to `arg`.
func applyTranspose_arityN(f, arg) -> transpose return type

/// Makes a differentiable function from the given `original`, `jvp`, and `vjp` functions.
func differentiableFunction_arityN(original, jvp, vjp)

/// Makes a linear function from the given `original` and `transpose` functions.
func linearFunction_arityN(original, transpose)
```

Includes a SILGen filecheck test for all of these.

### CODE REVIEW

**Review by dan-zheng — APPROVED** [review]

**Comment by marcrasi** [comment]
@swift-ci please smoke test

**Review by rxwei — APPROVED** [review]

**Comment by dan-zheng** [comment]
Merging to unblock progress.

File: include/swift/AST/KnownIdentifiers.def
Comment (davezarzycki) [inline include/swift/AST/KnownIdentifiers.def]: This list was alphabetized until this commit landed. If you have the time, can you please fix this? Thanks

Comment (dan-zheng) [reply include/swift/AST/KnownIdentifiers.def]: `KnownIdentifiers.def` organization will be improved in https://github.com/apple/swift/pull/30628, where I group together identifiers for differentiable programming in alphabetical order:
```diff
 // Differentiable programming
+IDENTIFIER(differential)
+IDENTIFIER(pullback)
 IDENTIFIER(TangentVector)
+IDENTIFIER(zero)
```

I hope that addresses your concern!

**Pull request merged**
